core/src/main/java/feign/FeignException.java

```diff
@@ -113,109 +146,113 @@ private static FeignException errorStatus(int status, String message, byte[] bod
       super(403, message, body);
     }
   }
 
-  public static class NotFound extends FeignException {
+  public static class NotFound extends FeignClientException {
     public NotFound(String message, byte[] body) {
       super(404, message, body);
     }
   }
 
-  public static class MethodNotAllowed extends FeignException {
+  public static class MethodNotAllowed extends FeignClientException {
     public MethodNotAllowed(String message, byte[] body) {
       super(405, message, body);
     }
   }
 
-  public static class NotAcceptable extends FeignException {
+  public static class NotAcceptable extends FeignClientException {
     public NotAcceptable(String message, byte[] body) {
       super(406, message, body);
     }
   }
 
-  public static class Conflict extends FeignException {
+  public static class Conflict extends FeignClientException {
     public Conflict(String message, byte[] body) {
       super(409, message, body);
     }
   }
 
-  public static class Gone extends FeignException {
+  public static class Gone extends FeignClientException {
     public Gone(String message, byte[] body) {
       super(410, message, body);
     }
   }
 
-  public static class UnsupportedMediaType extends FeignException {
+  public static class UnsupportedMediaType extends FeignClientException {
     public UnsupportedMediaType(String message, byte[] body) {
       super(415, message, body);
     }
   }
 
-  public static class TooManyRequests extends FeignException {
+  public static class TooManyRequests extends FeignClientException {
     public TooManyRequests(String message, byte[] body) {
       super(429, message, body);
     }
   }
 
-  public static class UnprocessableEntity extends FeignException {
+  public static class UnprocessableEntity extends FeignClientException {
     public UnprocessableEntity(String message, byte[] body) {
       super(422, message, body);
     }
   }
 
-  public static class InternalServerError extends FeignException {
+  public static class FeignServerException extends FeignException {
+    public FeignServerException(int status, String message, byte[] body) {
+      super(status, message, body);
+    }
+  }
+
+  public static class InternalServerError extends FeignServerException {
     public InternalServerError(String message, byte[] body) {
       super(500, message, body);
     }
   }
 
-  public static class NotImplemented extends FeignException {
+  public static class NotImplemented extends FeignServerException {
     public NotImplemented(String message, byte[] body) {
       super(501, message, body);
     }
   }
 
-  public static class BadGateway extends FeignException {
+  public static class BadGateway extends FeignServerException {
     public BadGateway(String message, byte[] body) {
       super(502, message, body);
     }
   }
 
-  public static class ServiceUnavailable extends FeignException {
+  public static class ServiceUnavailable extends FeignServerException {
```

Comment:
I think all 5xx errors would need to inherit from `RetryableException`. 

And, the `Retryer` defaults are repeat after 100ms, 150ms... up to a second. 5xx error might not be gone by then.

Is that fine?

Reply:
Not sure if I wanna retry on 500. Imagine server has an NPE and returning 500.

Unlikely that retrying would fix it.

Reply:
This PR is related to this comment: https://github.com/OpenFeign/feign/pull/825#discussion_r238952648 - a hierarchy of Client / Server Exceptions. 

I'm not sure what you meant - all 5xx inheriting from `FeignServerException` and 503 from `RetryableException`? 

I'm afraid we're going to need some other way than inheritance, to indicate that `ServiceUnavailable` should be retriable. 

I can think of an annotation or a simple list - what do you think?

Reply:
Only the 503 make sense to be retryable.

Which does break apart this whole inheritance for Server and Client exceptions.

May be we could remove the `RetryableException` (well, not really remove, just deprecate and stop using it on core) and add a different mechanism to describe if exception should be retry.  `@Annotation` seems over complicated. Maybe just a flag on `FeignException#shouldRetry`.

Reply:
I would think all server exceptions should be retryable. They just mean there was some issue on the server, but that the request wasn't bad, so if you retry after the server issue is fixed (maybe a dependent service was restarting) it might all be good again.

However, after looking at the list, I don't see how "501 - Not Implemented" or "505 - HTTP Version Not Supported" is a server exception (I'd think those should be 4xx bad requests).

Reply:
I think that we can table this discussion around retry and address it in another issue.